File: Graphs/23/time_needed_dfs.py
class Solution:
    def numOfMinutes(self, n: int, headID: int, manager: List[int], informTime: List[int]) -> int:
        if n <= 1:
            return 0
        subordinates = [[] for _ in range(len(manager))]
        for i, ele in enumerate(manager):
            if ele > -1:
                subordinates[ele].append(i)

        return self.dfs(headID, subordinates, informTime, 0, 0)

    def dfs(self, id, subordinates, informTime, minutes, maximum):
        if len(subordinates[id]) == 0: return maximum

        minutes = informTime[id] + minutes
        maximum = max(maximum, minutes)

        for ele in subordinates[id]:
            maximum = self.dfs(ele, subordinates, informTime, minutes, maximum)
        
        return maximum

File: Graphs/23/time_needed_dfs.js
var numOfMinutes = function (n, headID, manager, informTime) {
  if (n <= 1) return 0;
  const subordinates = manager.map(() => []);
  for (let i = 0; i < manager.length; i++) {
    if (manager[i] > -1) {
      subordinates[manager[i]].push(i);
    }
  }

  return dfs(headID, subordinates, informTime, 0, 0);
};

var dfs = function (id, subordinates, informTime, minutes, max) {
  // console.log(id);
  // console.log(subordinates[id], subordinates)
  if (subordinates[id].length === 0) {
    return max;
  }

  minutes = informTime[id] + minutes;
  max = Math.max(max, minutes);

  for (let i = 0; i < subordinates[id].length; i++) {
    max = dfs(subordinates[id][i], subordinates, informTime, minutes, max);
  }
  return max;
};
